File: python/forge3d/instancing.py
#!/usr/bin/env python3
"""A22: Instanced Geometry (PT) - TLAS-style instances"""

import numpy as np
from typing import List, Tuple

class InstancedGeometry:
    """A22: 10k instances with one BLAS; ≤512MiB VRAM."""

    def __init__(self, max_instances: int = 10000):
        self.max_instances = max_instances
        self.instances = []
        self.shared_blas = None  # Shared BLAS reference

    def add_instance(self, transform: np.ndarray, material_id: int = 0) -> int:
        """Add instance with per-instance transform."""
        if len(self.instances) >= self.max_instances:
            raise ValueError("Maximum instances exceeded")

        instance_data = {
            'transform': transform.copy(),
            'material_id': material_id,
            'blas_index': 0  # Shared BLAS
        }
        self.instances.append(instance_data)
        return len(self.instances) - 1

    def get_memory_usage(self) -> int:
        """Calculate memory usage in bytes."""
        # Each instance: 4x4 transform + metadata
        bytes_per_instance = 16 * 4 + 4 + 4  # transform + material + blas
        return len(self.instances) * bytes_per_instance

    def validate_memory_budget(self) -> bool:
        """A22: ≤512MiB VRAM requirement."""
        max_vram = 512 * 1024 * 1024  # 512 MiB
        return self.get_memory_usage() <= max_vram

File: src/accel/instancing.rs
//! A22: Instanced Geometry (PT) - TLAS-style instances with per-instance transforms

use glam::Mat4;
use wgpu::*;

#[repr(C)]
#[derive(Debug, Clone, Copy, bytemuck::Pod, bytemuck::Zeroable)]
pub struct InstanceData {
    pub transform: [f32; 16], // 4x4 matrix
    pub inv_transform: [f32; 16],
    pub blas_index: u32,
    pub material_id: u32,
    pub _padding: [u32; 2],
}

pub struct TLAS {
    instances: Vec<InstanceData>,
    max_instances: usize,
    #[allow(dead_code)]
    device: std::sync::Arc<Device>,
}

impl TLAS {
    pub fn new(device: std::sync::Arc<Device>, max_instances: usize) -> Self {
        Self {
            instances: Vec::new(),
            max_instances,
            device,
        }
    }

    // A22: 10k instances with one BLAS; ≤512MiB VRAM
    pub fn add_instance(
        &mut self,
        transform: Mat4,
        blas_index: u32,
        material_id: u32,
    ) -> Result<usize, String> {
        if self.instances.len() >= self.max_instances {
            return Err("Maximum instances exceeded".to_string());
        }

        let inv_transform = transform.inverse();

        let instance = InstanceData {
            transform: transform.to_cols_array(),
            inv_transform: inv_transform.to_cols_array(),
            blas_index,
            material_id,
            _padding: [0; 2],
        };

        self.instances.push(instance);
        Ok(self.instances.len() - 1)
    }

    pub fn get_memory_usage(&self) -> usize {
        self.instances.len() * std::mem::size_of::<InstanceData>()
    }

    pub fn validate_memory_budget(&self) -> bool {
        const MAX_VRAM: usize = 512 * 1024 * 1024; // 512 MiB
        self.get_memory_usage() <= MAX_VRAM
    }
}
